### PR TITLE
Add very basic S3 tests and fix some S3 bugs

### PR DESCRIPTION
**Pull request by alanking**
Addresses #124 
Addresses #132 
Addresses #285

All tests are passing. I thought this was going to be an S3 bug fixing PR until I discovered that #124 is not exclusive to S3.

The new S3 test suite was loosely based on the `test_delete_modes` structure for doing things.

Also added a testing library to reduce some duplicate code.

### CODE REVIEW

File: irods_capability_automated_ingest/irods/s3_bucket.py [hunk 1]
```diff
@@ -378,10 +380,41 @@ def upload_file(logger, session, meta, op, **options):
...
+        options["destRescName"] = resc_name
+
+    if b64_path_str is not None:
+        source_physical_fullpath = base64.b64decode(b64_path_str)
+
+    logger.info(
+        "syncing object %s, options = %s" % (dest_dataobj_logical_fullpath, options)
+    )
+
+    # TODO: Issue #208
```
Comment (korydraughn) [inline irods_capability_automated_ingest/irods/s3_bucket.py:408]: Please update the structure of this TODO.

File: irods_capability_automated_ingest/test/test_s3_bucket_scan.py [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+            **test_lib.get_test_irods_client_environment_dict()
+        )
+        cls.job_name = "test_s3_sync_job"
+        # TODO(#286): Derive this from the environment...
+        cls.s3_endpoint_domain = "minio:19000"
+        # TODO(#286): Derive these from the environment...
+        cls.s3_access_key = "irods"
+        cls.s3_secret_key = "irodsadmin"
+        f = tempfile.NamedTemporaryFile("w+t", delete=False)
+        # TODO(#???): This will not work on Windows...
```
Comment (korydraughn) [inline irods_capability_automated_ingest/test/test_s3_bucket_scan.py:99]: Need issue number for TODO.

File: irods_capability_automated_ingest/test/test_s3_bucket_scan.py [hunk 3]
```diff
@@ -409,6 +409,31 @@ def test_s3_with_register_as_replica_sync(self):
             finally:
                 self.minio_client.remove_object(self.bucket_name, new_object_name)
 
+    def test_register_to_deep_nonexistent_subcollection_does_not_hang_forever__issue_124(
+        self,
+    ):
+        operation = Operation.REGISTER_SYNC
+        event_handler_contents = test_s3_sync_operations.get_event_handler(operation)
+        # If the destination collection's parent collection does not exist, the bug is that the sync job will hang
+        # forever.
```
Comment (alanking) [inline irods_capability_automated_ingest/test/test_s3_bucket_scan.py:418]: This comment should probably read like this:

```suggestion
        # The destination collection needs to have enough path elements to exceed the number of path elements in
        # the "path" to the S3 object.
```

File: irods_capability_automated_ingest/irods/s3_bucket.py [hunk 4]
```diff
@@ -271,7 +273,7 @@ def register_file(hdlr_mod, logger, session, meta, **options):
     )
 
 
-def upload_file(logger, session, meta, op, **options):
+def _upload_file(logger, session, meta, op, exists=False, **options):
```
Comment (alanking) [inline irods_capability_automated_ingest/irods/s3_bucket.py:276]: I don't really care for this solution, but it gets the job done... perhaps a better name is all it needs...

File: irods_capability_automated_ingest/test/test_s3_bucket_scan.py [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+            **test_lib.get_test_irods_client_environment_dict()
+        )
+        cls.job_name = "test_s3_sync_job"
+        # TODO(#286): Derive this from the environment...
+        cls.s3_endpoint_domain = "minio:19000"
+        # TODO(#286): Derive these from the environment...
+        cls.s3_access_key = "irods"
+        cls.s3_secret_key = "irodsadmin"
+        f = tempfile.NamedTemporaryFile("w+t", delete=False)
+        # TODO(#???): This will not work on Windows...
```
Comment (alanking) [inline irods_capability_automated_ingest/test/test_s3_bucket_scan.py:99]: This isn't really like #250 since it's not a path, but it is a problem for Windows nonetheless. This will not be a problem when #264 is resolved.

**Review by korydraughn — COMMENTED** [review]
Looks good.

File: docker/ingest-test/docker-compose.yml [hunk 6]
```diff
@@ -55,5 +55,15 @@ services:
       - some-redis
       - irods-catalog-provider
 
+  minio:
```
Comment (trel) [inline docker/ingest-test/docker-compose.yml:58]: consider the updates included in

https://github.com/irods/irods_demo/blob/main/docker-compose.yml#L137-L149

File: irods_capability_automated_ingest/tasks/s3_bucket_tasks.py [hunk 7]
```diff
@@ -301,7 +301,7 @@ def s3_bucket_sync_entry(self, meta_input, datafunc, metafunc):
             # Construct S3 "logical path"
             target2 = "/".join([meta["target"], reg_path])
             # Construct S3 "physical path" as: /bucket/objectname
-            meta["path"] = f"/{meta['root']}/{path}"
+            meta2["path"] = f"/{meta['root']}/{path}"
```
Comment (trel) [inline irods_capability_automated_ingest/tasks/s3_bucket_tasks.py:304]: what ... is happening here?

Comment (alanking) [reply irods_capability_automated_ingest/tasks/s3_bucket_tasks.py:304]: This fixes the issue for S3 data uploading by modifying the correct copy of `meta` which is passed to the uploading function. `meta` is the input dict. `meta2` is a copy which is then modified to be passed to another call so that the original `meta` is not changed.

File: irods_capability_automated_ingest/test/test_s3_bucket_scan.py (hunk 3 above)
Comment (alanking) [reply irods_capability_automated_ingest/test/test_s3_bucket_scan.py]: Done

File: irods_capability_automated_ingest/test/test_s3_bucket_scan.py (hunk 5 above)
Comment (alanking) [reply irods_capability_automated_ingest/test/test_s3_bucket_scan.py]: Referencing #264.

File: irods_capability_automated_ingest/irods/s3_bucket.py (hunk 1 above)
Comment (alanking) [reply irods_capability_automated_ingest/irods/s3_bucket.py]: Done

File: irods_capability_automated_ingest/test/test_s3_bucket_scan.py (hunk 2 above)
Comment (alanking) [reply irods_capability_automated_ingest/test/test_s3_bucket_scan.py]: Referencing #264.

File: docker/ingest-test/docker-compose.yml (hunk 6 above)
Comment (alanking) [reply docker/ingest-test/docker-compose.yml:58]: Done

File: irods_capability_automated_ingest/tasks/s3_bucket_tasks.py (hunk 7 above)
Comment (trel) [reply irods_capability_automated_ingest/tasks/s3_bucket_tasks.py:304]: sold.   also, gah, horrible convention.   need better names eventually.

Comment (alanking) [reply irods_capability_automated_ingest/tasks/s3_bucket_tasks.py:304]: Yep. Resolving this, then.

File: irods_capability_automated_ingest/irods/s3_bucket.py (hunk 4 above)
Comment (alanking) [reply irods_capability_automated_ingest/irods/s3_bucket.py:276]: I feel that this is good enough for now. The underscore is pretty standard practice in Python to indicate that something should not be used by external modules, and the name does describe what happens in this function. Please open a new conversation if feelings are strong regarding the naming here.

File: docker/demo/compose.yaml [hunk 8]
```diff
@@ -46,14 +46,16 @@ services:
         command: ["-c", "2", "--loglevel", "INFO", "-n", "ingest-demo"] # Configure Celery options here. Note: Only takes effect at container creation.
 
     minio:
-        image: minio/minio:RELEASE.2024-06-28T09-06-49Z
-        command: server --address ":19000" --console-address ":19001" /data/minio-s3
+        image: minio/minio:RELEASE.2024-09-13T20-26-02Z
         ports:
             - "19000:19000" # This is the port to use for issuing S3 requests.
             - "19001:19001" # Change this port, if needed, to access the MinIO console webpage.
+        command: server --address ":19000" --console-address ":19001" /data/minio-s3
```
Comment (trel) [inline docker/demo/compose.yaml:53]: no need for address and console-address on this line due to ENV variables below

File: docker/ingest-test/docker-compose.yml [hunk 9]
```diff
@@ -56,14 +56,16 @@ services:
       - irods-catalog-provider
 
   minio:
-    image: minio/minio:RELEASE.2024-06-28T09-06-49Z
-    command: server --address ":19000" --console-address ":19001" /data/minio-s3
+    image: minio/minio:RELEASE.2024-09-13T20-26-02Z
     ports:
       - "19000:19000" # This is the port to use for issuing S3 requests.
       - "19001:19001" # Change this port, if needed, to access the MinIO console webpage.
+    command: server --address ":19000" --console-address ":19001" /data/minio-s3
```
Comment (trel) [inline docker/ingest-test/docker-compose.yml:63]: no need for address and console-address on this line due to ENV variables below

Comment (trel) [reply docker/ingest-test/docker-compose.yml:63]: `command: minio server /data/minio-s3`

File: docker/demo/compose.yaml (hunk 8 above)
Comment (trel) [reply docker/demo/compose.yaml:53]: `command: minio server /data/minio-s3`

File: docker/demo/compose.yaml [hunk 10]
```diff
@@ -50,7 +50,7 @@ services:
         ports:
             - "19000:19000" # This is the port to use for issuing S3 requests.
             - "19001:19001" # Change this port, if needed, to access the MinIO console webpage.
-        command: server --address ":19000" --console-address ":19001" /data/minio-s3
+        command: server /data/minio-s3
```
Comment (trel) [inline docker/demo/compose.yaml:53]: minio server /data/minio-s3 ?

File: docker/demo/compose.yaml (hunk 8 above)
Comment (alanking) [reply docker/demo/compose.yaml]: Done - seems to be working

File: docker/demo/compose.yaml (hunk 10 above)
Comment (alanking) [reply docker/demo/compose.yaml:53]: This is working because the entrypoint script used by the Minio Dockerfile prepends `minio` to the command if it doesn't start with that: https://github.com/minio/minio/blob/cbfe9de3e75f2475ac96cb3690ce2d39a4e63690/dockerscripts/docker-entrypoint.sh#L4-L9

We should not lean on this behavior, so I'll put `minio` at the front as you've suggested.

Comment (alanking) [reply docker/demo/compose.yaml]: Done

Comment (trel) [reply docker/demo/compose.yaml]: nice - thanks for looking that up.

**Comment by trel** [comment]
squash to taste.

**Comment by alanking** [comment]
Squashed

**Review by korydraughn — APPROVED** [review]
If the tests are passing, pound it.

**Comment by alanking** [comment]
#'d, mergin
